README.md
# SIXDB (simple indexedDB)  



![GitHub tag (latest by date)](https://img.shields.io/github/tag-date/jjcapellan/sixdb.svg)
![GitHub license](https://img.shields.io/github/license/jjcapellan/SIXDB.svg)


  

SIXDB is a basic and very easy to use library wrapper for indexedDB.  
IndexedDB is the recommended solution for the persistent storage of large volumes of structured data in the browser.  
But, IndexedDB lacks a query language, is asynchronous and can be complex to manage.  
SIXDB adds an abstraction layer over indexedDB that hides that complexity.  

You can learn how to use SIXDB in less than 5 minutes (or maybe 6 ;)).  
**Complete SIXDB documentation available [here](https://jjcapellan.github.io/SIXDB/)**


## In this document ...
***
* [Features](#Features)
* [Installation](#Installation)
* [Tutorial](#Tutorial)
    * [Database structure creation](#Database-structure-creation)
    * [Insert records](#Insert-records)
    * [Read records](#Read-records)
    * [Update records](#Update-records)
    * [Delete records](#Delete-records)
* [Query language](#Query-system)
* [Custom comparison operator](#Custom-operator)
* [Using functions as values](#Using-functions)
* [The task queue](#Task-queue)
* [The SIXDB error handling](#Error-handling)
* [License](#License)

## <a name="Features"></a>Features
***
* [Task queue](#Task-queue) system that manages all asynchronous operations.
* Natural and intuitive [Query language](#Query-system).
* Simple and flexible methods to perform CRUD operations.
* Join operations support.
* In addition to the predefined aggregate functions (sum, max, min, avg) you can use your own functions.
* Insertion operations allow several entries to be entered at the same time.
* Update operations accept [functions as value](#Using-functions) to modify the current value of the record.
* Lightweight. SIXDB takes less than 100Kb minified and less than 10Kb compressed with gzip.
* Default errorCallback in all methods. ErrorCallback parameter is optional.
* Complete [documentation](https://jjcapellan.github.io/SIXDB/) with examples.



## <a name="Installation"></a>Installation
***
There are two alternatives:
* Download the file [sixdb.js](https://cdn.jsdelivr.net/gh/jjcapellan/SIXDB@3.2.0/dist/sixdb.js) to your proyect folder and add a reference in your html:
```html
<script src = "sixdb.js"></script>
```
* Point a script tag to the CDN link:
```html
<script src = "https://cdn.jsdelivr.net/gh/jjcapellan/SIXDB@3.2.0/dist/sixdb.js"></script>
``` 


## <a name="Tutorial"></a>Tutorial
***
For the tutorial we will use an employees list from a supposed factory:

| ID | Name | Department    | Age | Salary
| --- |:--- | :--- | ---: | ---:
| 1 | Peter | manufacturing | 32 | 1200
| 2 | Paul | manufacturing | 27 | 1150
| 3 | Adam | manufacturing | 38 | 1260
| 4 | Alice | accounting | 41 | 1400
| 5 | Alex | manufacturing | 33 | 1250
| 6 | Kathy | manufacturing | 28 | 1150
| 7 | David | accounting | 32 | 1380

Each row represents a record. IndexedDB treats records as objects where each property represents a column.  
These records are stored in an **object store**. The **object store** accepts objects of different types.  
I could store other record without age column and with a lastName column.

The **index** is a subset of records from the object store that share the same property (**keypath**).  
The **index** order the records by its **keypath**.


### <a name="Database-structure-creation"></a>**Database structure creation**

```javascript
// First step is instantiate an SIXDB object, it checks if already exists a database called "companyDB" (If doesn't, a new one is created).
//   
const mydb = new Sixdb('companyDB');

// Now a store named "southFactory" with keyPath "id" and autoIncrement attribute activated is created.
// With this arguments, each new record added to store will receive a property "id" with a self-generated value.
// We need to create at least one store.
//
mydb.newStore(
    'southFactory',                            // Name of the new store.
    { keyPath: 'ID', autoIncrement: true }     // Optional object. Options are: keyPath, autoIncrement, successCallback, errorCallback.
);

// openStore() instantiates the new store. 
//
let myStore = mydb.openStore('southFactory');


//
// It's a good idea to create one or more indexes for each store to order the records (objects).
// The index keypath is the object property used to select and order the objects.
// Those objects that hasn't the index keypath as property will not be included in the index.
// The "unique" attribute prevents add two records with the same key. By default, "unique" is false.
//
// In this case we create a single index named "Names" with the keypath "name". 
//
myStore.newIndex(
    'Names',            // Name of the new index.
    'name',             // Keypath.
    { unique: false}    // Optional object. Options are: unique, succesCallback, errorCallback.
);



// ***** VERY IMPORTANT ****
// Once we have introduced the operations that we want to perform on the database, 
// we must use the Sixdb method execTasks() to execute them.
//
mydb.execTasks();
```


### <a name="Insert-records"></a>**Insert records**
With SIXDB we can insert records individually or through an objects array.

```javascript
//
// First step is instantiate a sixdb object, it checks if already exists a database called "companyDB" (If doesn't, a new one is created).
//   
const mydb = new Sixdb('companyDB');

// Opens "southFactory" object store
//
let myStore = mydb.openStore('southFactory');


// Inserts one record in "southFactory" object store.
//
myStore.add(
    {name: 'Peter', department: 'manufacturing', age: 32, salary: 1200}    // A single object that represents a record.
);


//
// Inserts 6 records in "southFactory" object store using an objects array.
//
myStore.add(
    'southFactory',
    [
        { name: 'Paul', department: 'manufacturing', age: 27, salary: 1150},
        { name: 'Adam', department: 'manufacturing', age: 38, salary: 1260},
        { name: 'Alice', department: 'accounting', age: 41, salary: 1400},
        { name: 'Alex', department: 'manufacturing', age: 33, salary: 1250},
        { name: 'Kathy', department: 'manufacturing', age: 28, salary: 1150},
        { name: 'David', department: 'accounting', age: 32, salary: 1380}
    ]
);


// ***** VERY IMPORTANT ****
// Once we have introduced the operations that we want to perform on the database, 
// we must use the Sixdb method execTasks() to execute them.
//
mydb.execTasks();
```

### <a name="Read-records"></a>**Read records**
SIXDB can read the records using the Store or the Index objects.  
The Index sorts the results by its keypath and the Store uses its primary key.  
Store and Index objects have the methods **get()** and **getAll()**.

```javascript
//
// First step is instantiate an sixdb object, it checks if already exists a database called "companyDB" (If doesn't, a new one is created).
//   
const mydb = new Sixdb('companyDB');

// Opens "southFactory" store
//
let myStore = mydb.openStore('southFactory');

// Opens "Names" index
//
let myIndex = myStore.openIndex('Names');


// Gets all records in the store. Gets the records sorted by the store primary key ("ID").
myStore.getAll(succesCallback);

// Gets all records in the index. Gets the records sorted by the index keypath ("name").
myIndex.getAll(successCallback);

// Gets the record with primary key ("ID") 4 from the store. Single value in first parameter refers to the keypath.
//
myStore.get(
    4,                 // The query.
    successCallback    // Success callback wich receives the result.
);

// Gets records from manufacturing department with salary > 1200.
//
myStore.get('department = manufacturing & salary > 1200', successCallback);


// ***** VERY IMPORTANT ****
// Once we have introduced the operations that we want to perform on the database, 
// we must use the Sixdb method execTasks() to execute them.
//
mydb.execTasks();


//
// Simple function that receives the uery result from SIXDB methods and show it on console.
// The result received can be a single object or an object array. 
// 
//
function successCallback(result) {
    if(Array.isArray(result)){
        for (let i = 0, j = result.length; i < j; i++) {
            console.log(
             `Name: ${result[i].name}\n` +
             `Department: ${result[i].department}\n` + 
             `Age: ${result[i].age}\n` +
             `Salary: ${result[i].salary}\n`
             );
        }
    } else {
        console.log(
         `Name: ${result[i].name}\n` +
         `Department: ${result[i].department}\n` + 
         `Age: ${result[i].age}\n` +
         `Salary: ${result[i].salary}\n`
         );
    }
}
```

### <a name="Update-records"></a>**Update records**
The Store method **update()** allows us to update the records in many ways.
```javascript
//
// First step is instantiate an sixdb object, it checks if already exists a database called "companyDB" (If doesn't, a new one is created).
//   
const mydb = new Sixdb('companyDB');

// Opens "southFactory" store
//
let myStore = mydb.openStore('southFactory');

// This simple example changes the salary and age of the employee with ID 2.
//
myStore.update(
    2,                                  // The query.
    {salary: 1290, age: 39},            // Object with the new values.
    {errorCallback: myErrorCallback}    // Optional object. Options are: successCallback, errorCallback.
);

// Updates the salary of all records where department is manufacturing and age > 30 to a new salary of 1300.
//
myStore.update(
    'department = "manufacturing" & age > 30',    // The query
    {salary: 1300},                               // Object with the new values
    {errorCallback: myErrorCallback }             // Optional object.
);

// The object values parameter accepts functions as value whose receives the old value an returns the new value.
// Here all salaries from employees with age highter than 40 or with salary less than 600, are increased by 300.
//
myStore.update(
    'age > 40 | salary < 600',
    {salary: function(oldSalary) { return oldSalary + 300;} },           // A function is used as new value for salary.
    {errorCallback: myErrorCallback}
);

// Simple error callback
//
function myErrorCallback(e){
    console.error(e);
};


// ***** VERY IMPORTANT ****
// Once we have introduced the operations that we want to perform on the database, 
// we must use the Sixdb method execTasks() to execute them.
//
mydb.execTasks();
```

### <a name="Delete-records"></a>**Delete records**
Store object uses the method **del()** for delete one or more records.
```javascript
//
// First step is instantiate an sixdb object, it checks if already exists a database called "companyDB" (If doesn't, a new one is created).
//   
const mydb = new Sixdb('companyDB');

// Opens "southFactory" store
//
let myStore = mydb.openStore('southFactory');

// Deletes the record with ID 5 from the object store "southFactory".
//
myStore.del(
    5,                                  // Query. A single value refers to the primary key.
    {errorCallback: myErrorCallback}    // Optional object. Options are: successCallback, errorCallback.            
);

// deletes all records from manufacturing department with salaries highter than 1200 or age highter than 60
//
myStore.del(
    '(department = manufacturing) & (salary > 1200 | age > 60)'             // Query with 2 groups of conditions
);

// ***** VERY IMPORTANT ****
// Once we have introduced the operations that we want to perform on the database, 
// we must use the Sixdb method execTasks() to execute them.
//
mydb.execTasks();
```



## <a name="Query-system"></a>**Query language**
***
Some SIXDB methods receive as parameter to select records a string with an expression that represents the query.
Write a query is very intuitive. Is similar to write the condition in an "if" sentence.  
A simple query would be: `recordProperty = value` 
There are some rules:
* You can use these normal comparisson operators:  **=** , **>**, **\<**, **>=**, **<=**, **!=**
* There are too "special" operators:
  * **<>**   : "Contains" operator. Valid for strings.
  ```javascript
  name <> ul // The property name contains the substring "ul". (quotes are optional)
  ```
  * **^**    : "Starts with" operator. Valid for strings. Is case sensitive.
  ```javascript
  name ^ Pe // Means the property name starts with the substring "Pe" (quotes are optional).
  ```
* The valid logical operators are: **&**, **&&**, **|**, **||**
* **&** ("and") has same effect than **&&**, and the same to **|** ("or") and **||**.
* Only one type of logical operator can be used in a group of conditions. 
```javascript
salary > 1000 & name != 'Peter' & age > 34      // This is correct
salary > 1000 & name != 'Peter' | age > 34      // This is wrong
```
* Groups of conditions within other groups are not allowed.
```javascript
(salary > 1000 & name != 'Peter') | (name = 'Adam' & salary < 900)            // This is correct
((salary > 1000 & name != 'Peter') | ID = 5) | (name = 'Adam' & salary < 900) // This is wrong
```
* Outside of a conditions group there can't be another without parentheses.
```javascript
(salary > 1000 & name != 'Peter') | (name = 'Adam')   // This is correct
(salary > 1000 & name != 'Peter') | name = 'Adam'     // This is wrong
```
* Only one type of logical operator can be used between groups of conditions
```javascript
(a = 2 & c < 10) | (d != 1) | (d = 10 & e >= 12)  // This is correct
(a = 2 & c < 10) | (d != 1) & (d = 10 & e >= 12)  // This is wrong
```
* A single value refers to a keypath
```javascript
myStore.del(
    5,                                  // Query. A single value refers to the primary key.
    {errorCallback: myErrorCallback}    // Optional object. Options are: successCallback, errorCallback.            
);
```
* Spaces or symbols in the value of a condition must be enclosed in quotation marks. Parentheses and nested quotes are not supported.
```javascript
name = 'John Smith'                 // This is correct
equation = 'e = v / t'              // This is correct
equation = e = v / t                // This is wrong. The query system can't parse that "=" without quotes.
message = 'This is my "message"'    // This is wrong.
message = 'This (is) my message'    // This is wrong.
```

## <a name="Custom-operator"></a>**Custom comparison operator**
***
We can create our own operator to use in the queries.  
The custom operator is represented in the query system with "\~\~".   
To get it we must make a compare function and set the customOperator variable.  
By default "\~\~" acts like "=".
```javascript
//
// The compare function must have two arguments, property value and test value. If this function triggers
// an error exception then the query system returns the condition result as false.
// In this example the function returns true when the length of property and test value is equal.
//
const compareFunction = function(propValue, testValue){
    return propValue.length == testValue.length;
};

// Assigns compareFunction to the customOperator variable
//
mydb.setCustomOperator(compareFunction);
``` 


## <a name="Using-functions"></a>**Using functions as values**
***
SIXDB allows send a function in a query to modify records value.  
To do this, we will consider the function as if it were the new value in the update.records() method.  
This function will receive the current record value as a parameter and return the new value.

```javascript
// The object values parameter accepts functions as value whose receives the old value an returns the new value.
// Here all salaries from employees with age highter than 40 or with salary less than 600, are increased by 300.
//
myStore.update(
    'age > 40 | salary < 600',
    {salary: function(oldSalary) { return oldSalary + 300;} },           // A function is used as new value for salary.
    {errorCallback: myErrorCallback}
);
```

## <a name="Task-queue"></a>**The SIXDB task queue**  
***

CHANGELOG.MD
## v3.3.1
* Fix: null callback parammeters cause errors in several functions.
## v3.3.0
* Added function Sixdb.checkStore().
* Added function Store.checkIndex().
## v3.2.1
* Fix: Store.newIndex() disabled.
## v3.2.0
* Added function Sixdb.join(). Add basic support to join operations.
* Fix: Store.newIndex() close connection when index exists.
## v3.1.0
* Added function Store.clear(). Deletes all records in the store.
* Added getters Store.name(), Index.name(), Index.storeName().
* Fix: New instance of Index modifies indexName and storeName of previous instances.
* Fix: New instance of Store modifies storeName of previous instances.
## v3.0.1
* Fix: unique attribute unassigned in Store.newIndex().
* Fix: consoleOff self assigned in Sixdb.setConsoleOff().
## v3.0.0
* Redesigned api. Now databases, stores and indexes are objects with its own methods.
## v2.0.1
* Fix: succesCallback null in cursorLoop().
## v2.0.0
### New features
* Added function get.aggregateFn(). Applies an aggregate function to a property.  
Accepts as arguments predefined functions:  
  * aggregateFuncs.sum. Sums the values of a property.
  * aggregateFuncs.avg. Calculates the average of a property.
  * aggregateFuncs.max. Maximum value of a property.
  * aggregateFuncs.min. Minimum value of a property
* Custom aggregate functions. get.aggregateFn() accepts custom functions.
* Added function get.count(). Returns number of records of a query.
### Minor improvements and fixes
* Optional arguments now are grouped in an object.
* Default values in some arguments to avoid null errors.
* Less size of the minified file.
## v1.9.0
* Added function get.sum(). Sums the values of a property.
* Added function get.avg(). Calculates the average of a property.
* Query is optional in get.records() function.
## v1.8.0
* Added string comparison operator **$** ("ends with") to use in queries.
* Added custom comparison operator **\~\~** wich works with a custom compare function.
## v1.7.0
### New features
* Added string comparison operator **<>** ("contains") to use in queries.
* Added string comparison operator **^** ("starts with") to use in queries.
### Minor improvements and fixes
* DOMexceptions triggered by indexedDB method createdObjectStore now are correctly managed by a trycatch.
* Some refactoring that saves 5Kb to the file.
## v1.6.0
### New features
* Added an improved error handling system. Now detects and hanle more possible errors.
### Minor improvements and fixes
* Error notifications are not affected by the flag consoleOff.
* All undefined error callbacks are transformed in empty functions. This prevents a problem in the try catch helpers.

src/sixdb.js
/**
 * @author       Juan Jose Capellan <soycape@hotmail.com>
 * @copyright    2018 Juan Jose Capellan
 * @license      {@link https://github.com/jjcapellan/SIXDB/blob/master/LICENSE | MIT license}
 */

import {
  _qrySys,
  done,
  execTasks,
  logEnum,
  logger,
  requestErrorAction,
  requestSuccessAction,
  Store,
  tasks,
} from './index.js';


let db = null;
let dbName;
const voidFn = function () {
  return 0;
};

// Opens the database and stores the result in db
function openDb() {
  let request = window.indexedDB.open(dbName);

  request.onerror = function () {
    alert('Error. You must allow web app to use indexedDB.');
  };

  request.onsuccess = function (event) {
    db = event.target.result;
    done();
  };
}

// Predefined task to open the actual database
const tkOpen = { args: null, fn: openDb };

// Allows other modules modify variable db
function setDb(_db) {
  db = _db;
}

function checkStore(storeName, callback) {
  let origin = 'Sixdb.checkStore()';
  logger(origin + logEnum.begin);
  let exists = db.objectStoreNames.contains(storeName);
  db.close();
  callback(exists, origin);
  done();
  return;
}

// Creates a store in the database
function newStore(
  storeName,
  { keyPath, autoIncrement, successCallback, errorCallback } = {}
) {
  let version;
  let origin = 'Sixdb.newStore()';
  logger(origin + logEnum.begin);

  // If store already exist then returns
  if (db.objectStoreNames.contains(storeName)) {
    db.close();
    logger(`Object store "${storeName}" already exists`);
    done();
    return;
  }

  version = db.version;
  db.close();
  let newVersion = version + 1;
  let _store;

  let request = window.indexedDB.open(dbName, newVersion);

  request.onupgradeneeded = function (event) {
    db = event.target.result;

    try {
      _store = db.createObjectStore(storeName, {
        keyPath: keyPath,
        autoIncrement: autoIncrement
      });
    } catch (e) {
      requestErrorAction(origin, e, errorCallback);
      return;
    }

    _store.onerror = function (event) {
      requestErrorAction(origin, event.target.error, errorCallback);
    };
  };

  request.onsuccess = function (event) {
    requestSuccessAction(
      event,
      origin,
      successCallback,
      `New object store "${storeName}" created`
    );
  };
}

function delStore(storeName, { successCallback, errorCallback }) {
  let origin = 'Sixdb.delStore()';
  logger(origin + logEnum.begin);

  //// Gets the new version
  //
  let version = db.version;
  db.close();
  let newVersion = version + 1;

  //// The change of the database schema only can be performed in the onupgradedneeded event
  //// so a new version number is needed to trigger that event.
  //
  let request = window.indexedDB.open(dbName, newVersion);

  request.onupgradeneeded = function (event) {
    db = event.target.result;
    db.deleteObjectStore(storeName);
  };

  request.onsuccess = function (event) {
    requestSuccessAction(
      event,
      origin,
      successCallback,
      `Object store "${storeName}" deleted`
    );
  };

  request.onerror = function () {
    requestErrorAction(origin, request.error, errorCallback);
  };
}

function delDB({ successCallback, errorCallback }) {
  let origin = 'Sixdb.destroy()';
  logger(origin + logEnum.begin);

  let request = window.indexedDB.deleteDatabase(dbName);

  request.onerror = function () {
    requestErrorAction(origin, request.error, errorCallback);
  };

  request.onsuccess = function (event) {
    successCallback(event, origin);
    logger(`Database "${dbName}" deleted`);
    done();
  };
}

// Join operation. (ON store1.primaryKey = store2.index.keyPath)
function join({ store1Name, store2Name, indexName, succesCallback, errorCallback }) {
  let joinResult = [];
  let store1Result = [];
  let store2Result = [];
  let origin = 'getJoin()';
  logger(origin + logEnum.begin);
  let transaction = db.transaction([store1Name, store2Name]);
  let indexKeyPath;
  let posCursor1 = 0;
  let posCursor2 = 0;
  let store1ResultLength;
  let store2ResultLength;

  let store1 = transaction.objectStore(store1Name);
  let store1KeyPath = store1.keyPath;
  store1.getAll().onsuccess = function (event) {
    store1Result = event.target.result;
    store1ResultLength = store1Result.length;

    let store2 = transaction.objectStore(store2Name);
    let store2Index = store2.index(indexName);
    store2Index.getAll().onsuccess = function (event) {
      store2Result = event.target.result;
      store2ResultLength = store2Result.length;
      console.log(store2ResultLength);
      indexKeyPath = store2Index.keyPath;
      makeJoinResult();
    };
  };

  function makeJoinResult() {
    if (
      store1Result[posCursor1][store1KeyPath] == store2Result[posCursor2][indexKeyPath]
    ) {
      joinResult.push(Object.assign(store2Result[posCursor2], store1Result[posCursor1]));
      posCursor2++;
    } else {
      posCursor1++;
    }

    if (posCursor1 == store1ResultLength || posCursor2 == store2ResultLength) {
      requestSuccessAction(
        joinResult,
        origin,
        successCallback,
        'Join operation completed'
      );
    } else {
      makeJoinResult();
    }
  }
}

/**
 * Constructs a Sixdb instance.
 * @class
 * @param  {string} _dbName Name of the database 
 */
window.Sixdb = function (_dbName) {
  dbName = _dbName;

  // Query system from qrySys.js
  let qrySys = _qrySys;

  // Creates or opens the database
  function newDB(errorCallback = voidFn) {
    let request = window.indexedDB.open(dbName);
    let origin = 'Sixdb.newDB()';
    logger(origin + logEnum.begin);

    // Boolean: Database doesn't exist
    let noDb = false;

    // if onupgradeneeded means is a new database
    request.onupgradeneeded = function () {
      noDb = true;
    };

    request.onsuccess = function (event) {
      let db = event.target.result;
      db.close();
      if (noDb) {
        logger(`Database "${dbName}" created`);
      } else {
        logger(`Database "${dbName}" already exists`);
      }
      done();
    };
  }

  // Adds newDB function to the task queue
  function addDB(errorCallback) {
    let args = [errorCallback];
    let task = {
      args: args,
      fn: newDB
    };
    tasks.push(task);
  }

  //// public functions declaration /////////////////
  this.name;
  this.setConsoleOff;
  this.customTask;
  this.aggregateFuncs;
  this.execTasks;
  this.setCustomOperator;
  this.checkStore;
  this.newStore;
  this.openStore;
  this.delStore;
  this.destroy;

  //// Initialization ///////////////////////////////
  qrySys.init();
  addDB();
  execTasks();
};

/**
 * Gets the database name.
 * @method window.Sixdb#name
 * @instance
 * @return Name of the database
 */
Sixdb.prototype.name = function () {
  return dbName;
};

/**
 * Sets the consoleOff value.
 * @param  {boolean} _consoleOff If true, the console output is off and only errors appear in console.
 */
Sixdb.prototype.setConsoleOff = function (_consoleOff) {
  consoleOff = _consoleOff;
};

/**
 * Add a specific function to the Sixdb task queue.
 * @method window.Sixdb#customTask
 * @instance
 * @param  {function} fn Our custom function that we want to add to the task queue.
 * @param  {object} context Usually the keyword "this"
 * @param  {...any} args Arguments for the function.
 * @example
 * var mydb = new Sixdb('companyDB');
 * 
 * // Creates new store
 * mydb.newStore('southFactory');
 * 
 * // Opens the store in a variable
 * var store = mydb.openStore('southFactory');
 *
 *
 * // Inserts one record in "southFactory" object store.
 * //
 * store.add(
 *    {ID: 1, name: 'Peter', department: 'manufacturing', age: 32, salary: 1200}
 * );
 *
 *
 * //
 * // To add an own function to the task queue, the method  customTask() of the Sixdb object is used
 * //
 * //     customTask( fn, context, args)
 * //
 * // This task is executed after the previous insertion task and before the next reading task.
 * //
 * mydb.customTask(
 *    function(m){                                // Custom function
 *        alert(m);
 *    },
 *    this,                                       // Context. Usually "this".
 *    'Inserting operation finished !!'           // Arguments of the function. Can be a variable number of arguments.
 * );
 *
 *
 *
 * // Reads all records from "southFactory" object store.
 * //
 * store.getAll(mySuccessCallback);
 *
 *
 * // ***** VERY IMPORTANT ****
 * // Once we have introduced the operations that we want to perform on the database,
 * // we must use the function execTasks() to execute them.
 * //
 * mydb.execTasks();
 */
Sixdb.prototype.customTask = function (fn, context, args) {
  let argsArray = [];
  if (args) {
    for (let i = 2, j = arguments.length; i < j; i++) {
      argsArray[2 - i] = arguments[i];
    }
  }
  let task = { type: 'custom', fn: fn, context: context, args: argsArray };

  tasks.push(task);
};
/** 
 * Contains predefined aggregate functions to use in aggregateFn method.<br>
 * The method aggregateFn accepts too custom functions.
 * 
 * @memberof window.Sixdb
 * @namespace
*/
Sixdb.prototype.aggregateFuncs = {
  /**
   * Sums two values
   * @param  {string | number} actual Acumulated value
   * @param  {string | number} selected Selected value
   * @return {string | number} Returns the sum.
   */
  sum(actual, selected) {
    return actual + selected;
  },

  /**
   * Calculates the average value.
   * @param  {number} actual Acumulated value.
   * @param  {number} selected Selected value.
   * @param  {number} counter Number of elements.
   * @return {number} average value.
   */
  avg(actual, selected, counter) {
    return (actual * (counter - 1) + selected) / counter;
  },

  /**
   * Returns the maximum value.
   * @param  {string | number} actual 
   * @param  {string | number} selected 
   * @return {string | number} The maximum value.
   */
  max(actual, selected) {
    return selected > actual ? selected : actual;
  },

  /**
   * Returns the minimum value.
   * @param  {string | number} actual 
   * @param  {string | number} selected 
   * @param  {number} counter Number of iteration.
   * @return {string | number} The minimum value.
   */
  min(actual, selected, counter) {
    if (counter == 1) {
      // First value of actual is null. Without this, min is allways null
      actual = selected;
    }
    return selected < actual && counter > 1 ? selected : actual;
  }
};

/**
   * Sets customOperator. To make the queries we can add to the Sixdb comparison operators our own operator.<br>
   * This operator will be represented by <b>~~</b>.
   * @method window.Sixdb#setCustomOperator
   * @instance
   * @param  {function} compareFunction Function to compare a property value with a test value.<br>
   * @example
   * var mydb = new Sixdb('myDatabase');
   *
   * //
   * // The compare function must have two arguments, property value and test value. If this function triggers
   * // an error exception, then the query system returns the condition as false.
   * //
   * mydb.setCustomOperator(
   *     function(propertyValue, testValue){
   *         return (propertyValue.length == testValue.length);
   *     });
   *
   */
Sixdb.prototype.setCustomOperator = function (compareFunction) {
  if (compareFunction) {
    if (typeof compareFunction == 'function') {
      if (compareFunction.length == 2) {
        customOperator = compareFunction;
      }
    }
  }
};

/**
   * Execs pending tasks. The tasks are executed sequentially.
   * A task does not run until the previous one ends.
   * <br>This avoids problems arising from the asynchronous nature of the indexedDB api.
   * @method window.Sixdb#execTask
   * @instance
   */
Sixdb.prototype.execTasks = function () {
  execTasks();
};

/**
 * Gets all records from the object store.
 * @method window.Sixdb#checkStore
 * @instance
 * @param {string} storeName Name of the store 
 * @param  {function} callback Receives a boolean (true if store exists in the database) and origin as parameters. 
 * @example
 * mydb.checkStore('Products', (exists) => {
 *  if(exists){
 *   console.log('Store Products exists in this database')
 *  }
 * });
 */
Sixdb.prototype.checkStore = function (storeName, callback = voidFn) {
  let args = [
    storeName,
    callback
  ];
  let task = {
    args: args,
    fn: checkStore
  };
  tasks.push(tkOpen);
  tasks.push(task);
};

/**
 * Creates a task wich creates a store object in the database.
 * @method window.Sixdb#newStore
 * @instance
 * @param  {string} storeName The store name.
 * @param  {object} [options]
 * @param  {string} [options.keyPath] The key path to be used by the new object store. 
 * <br>If empty or not specified, the object store is created without a key path and uses out-of-line keys. 
 * <br>You can also pass in an array as a keyPath.
 * @param  {Boolean} [options.autoIncrement] If true, the object store has a key generator. Defaults to false.
 * @param  {function} [options.succesCallback] Function called on success. Receives as parameters event and origin.
 * @param  {function} [options.errorCallback] Optional function to handle errors. Receives an error object as argument.
 */
Sixdb.prototype.newStore = function (
  storeName,
  { keyPath, autoIncrement, successCallback = voidFn, errorCallback = voidFn } = {}
) {
  let args = [
    storeName,
    {
      keyPath: keyPath,
      autoIncrement: autoIncrement,
      successCallback: successCallback,
      errorCallback: errorCallback
    }
  ];
  let task = {
    args: args,
    fn: newStore
  };
  tasks.push(tkOpen);
  tasks.push(task);
};

/**
 * Returns an existing store object ready to use
 * @method window.Sixdb#openStore
 * @instance
 * @param  {string} storeName Name of the store.
 * @return {object}
 */
Sixdb.prototype.openStore = function (storeName) {
  return new Store(storeName);
};

/**
 * Deletes an object store.
 * @method window.Sixdb#delStore
 * @instance
 * @param  {string} storeName Name of the object store. 
 * @param  {object} options
 * @param  {function} [options.succesCallback] Function called on success. Receives as parameters event and origin.
 * @param  {function} [options.errorCallback] Optional function to handle errors. Receives an error object as argument.
 */
Sixdb.prototype.delStore = function (
  storeName,
  { successCallback = voidFn, errorCallback = voidFn } = {}
) {
  let options = {
    successCallback: successCallback,
    errorCallback: errorCallback
  };
  let args = [storeName, options];
  let task = {
    args: args,
    fn: delStore
  };

  tasks.push(tkOpen);
  tasks.push(task);
};

/**
 * The current database is deleted.
 * @method window.Sixdb#destroy
 * @instance
 * @param  {object} options
 * @param  {function} [options.succesCallback] Function called on success. Receives as parameters event and origin.
 * @param  {function} [options.errorCallback] Optional function to handle errors. Receives an error object as argument.
 */
Sixdb.prototype.destroy = function (
  { successCallback = voidFn, errorCallback = voidFn } = {}
) {
  let options = {
    successCallback: successCallback,
    errorCallback: errorCallback
  };

  let args = [options];
  let task = {
    args: args,
    fn: delDB
  };

  tasks.push(task);
};

/**
 * Creates a join operation on two stores. Joins those objects where store1.keypath = store2.index.keypath <br>
 * and returns the result to a success callback.
 * @method window.Sixdb#join
 * @instance
 * @param  {object} options
 * @param  {string} options.store1Name Name of the store with a primary unique key
 * @param  {string} options.store2Name Name of the second store
 * @param  {string} options.indexName Name of the second store index
 * @param  {function} options.succesCallback Function called on success. Receives as parameters the join result and origin.
 * @param  {function} [options.errorCallback] Optional function to handle errors. Receives an error object as argument.
 * @example
 * // An example of object stored in store "Employees"
 * //
 * let person = {
 *     id: 1,                // <<<<<< keyPath of store1
 *     name: 'Peter',
 *     age: 32,
 *     salary: 1100
 * };
 * // An example of object stored in store "Productions"
 * //
 * let report = {
 *     reportId: 5,                
 *     employeeId: 1,        // <<<<<<<<<< keyPath of index "employeeIds" in store "Productions"
 *     production: 150
 * };
 * 
 * const mydb = new Sixdb('myDatabase');
 * 
 * mydb.join(
 * {
 * store1Name: 'Employees',
 * store2Name: 'Productions',
 * indexName: 'employeeIds',
 * successCallback: mySuccessCallback
 * }
 * );
 * 
 * // The expected results array received by the successCallback contains objects like:
 * // {id:1, name:'Peter, age:32, salary:1100, reportId:5, employeeId:1, production:150}   <<<<<<< id = employeeId
 * 
 *  
 * mydb.execTasks();
 */
Sixdb.prototype.join = function ({
  store1Name,
  store2Name,
  indexName,
  succesCallback = voidFn,
  errorCallback = voidFn
}) {
  let args = [
    {
      store1Name,
      store2Name,
      indexName,
      succesCallback,
      errorCallback
    }
  ];

  let task = {
    args: args,
    fn: join
  };

  tasks.push(tkOpen);
  tasks.push(task);
};

export { /*consoleOff,*/ db, newStore, dbName, tkOpen, setDb, voidFn };
